File: codeforces_2034C.go
package main

import (
	"fmt"
	"strconv"
	"bufio"
	"os"
	"strings"
)

func move(r, c int, mv rune) []int {
	switch mv {
	case 'U':
		return []int{r-1, c}
	case 'L':
		return []int{r, c-1}
	case 'D':
		return []int{r+1, c}
	case 'R':
		return []int{r, c+1}
	}

	panic("invalid move")
}

func main() {
	s := bufio.NewScanner(os.Stdin)
	s.Scan()
	t, err := strconv.Atoi(s.Text())
	if err != nil {
		panic(err)
	}

	for range t {
		s.Scan()
		nm := strings.Fields(s.Text())
		n, _ := strconv.Atoi(nm[0])
		m, _ := strconv.Atoi(nm[1])

		rows := []string{}
		for range n {
			s.Scan()
			row := s.Text()
			rows = append(rows, row)
		}

		cache := map[int]map[int]bool{}
		visited := map[int]map[int]bool{}
		var badCell func(r, c int) bool
		badCell = func(r, c int) bool {
			if cache[r] == nil {
				cache[r] = map[int]bool{}
			}

			if visited[r] == nil {
				visited[r] = map[int]bool{}
			}

			if result, ok := cache[r][c]; ok {
				return result
			}

			if visited[r][c] {
				cache[r][c] = true
				return true
			}

			if r < 0 || c < 0 || r >= n || c >= m {
				cache[r][c] = false
				return false
			}

			visited[r][c] = true
			defer func() { visited[r][c] = false }()

			if rows[r][c] != '?' {
				next := move(r,c,rune(rows[r][c]))
				nr := next[0]
				nc := next[1]

				result := badCell(nr, nc)

				cache[r][c] = result
				return result
			}

			for _, mv := range "URDL" {
				next := move(r,c,mv)
				nr := next[0]
				nc := next[1]

				isBad := badCell(nr, nc)

				if isBad {
					cache[r][c] = true
					return true
				}
			}

			cache[r][c] = false
			return false
		}

		count := 0
		for r := range n {
			for c := range m {
				if badCell(r, c) {
					count += 1
				}
			}
		}

		fmt.Println(count)
	}
}

File: codeforces_2034C.py
"""
This solution hits a TLE on code forces. The same logic implemented in golang passes all tests.
Not sure if there is anything wrong with the code that makes it slow
"""

import sys
sys.setrecursionlimit(1000000000)

def move(r, c, mv):
    if mv == 'U':
        return (r-1, c)
    if mv == 'R':
        return (r, c+1)
    if mv == 'D':
        return (r+1, c)
    if mv == 'L':
        return (r, c-1)

    raise('invalid move')


for _ in range(int(input())):
    n, m = map(int, input().split())
    ri = [[c for c in input()] for _ in range(n)]
    assert(m == len(ri[0]))

    cache = {}
    def bad_cell(r, c, visited):
        stack = [(r,c)]
        if (r,c) in cache:
            return cache[(r,c)]

        if (r,c) in visited:
            cache[(r,c)] = True
            return True

        if r < 0 or c < 0 or r >= n or c >= m:
            cache[(r,c)] = False
            return False

        if ri[r][c] != '?':
            nr, nc = move(r,c,ri[r][c])

            visited.add((r,c))
            result = bad_cell(nr, nc, visited)
            visited.remove((r,c))

            cache[(r,c)] = result
            return result

        for mv in 'URDL':
            nr, nc = move(r,c,mv)

            visited.add((r,c))
            isbad = bad_cell(nr, nc, visited)
            visited.remove((r,c))

            if isbad:
                cache[(r,c)] = True
                return True

        cache[(r,c)] = False
        return False

    count = 0
    for r in range(n):
        for c in range(m):
            if bad_cell(r, c, set()):
                count += 1

    print(count)
